uast/libuast/api.go

```diff
@@ -0,0 +1,278 @@
+	u.iface = iface
+	u.handle = C.uint64_t(h)
+	u.root = 0
+	u.ctx = 0
+	return u
+}
+
+//export UastNew
+// UastNew initializes a new UAST context that will use a provided node interface as an implementation.
+// This allows libuast to work with every language's native node data structures. Client can pass
+// an additional UastHandle to distinguish between different UAST context instances.
+//
+// The returned context pointer is guaranteed to be not NULL. Client should check LastError before
+// using the context and deallocate it with UastFree in case of an error occurs, or when the context
+// is no longer needed.
+func UastNew(iface *C.NodeIface, ctx C.UastHandle) *C.Uast {
+	cn := &cNodes{
+		impl: iface,
+	}
+	c := newContext(cn)
+
+	u := newUast(iface, c.Handle())
+	u.ctx = ctx
+	cn.ctx = u
+	return u
+}
+
+//export UastDecode
+// UastDecode accepts a pointer to a buffer with a specified size and decodes the content into
+// a new UAST structure.
+//
+// The new UAST context will use internal node interface implementation and all the nodes will
+// be managed by libuast.
+//
+// The returned context pointer is guaranteed to be not NULL. Client should check LastError before
+// using the context and deallocate it with UastFree in case of an error occurs, or when the context
+// is no longer needed.
+func UastDecode(p unsafe.Pointer, sz C.size_t, format C.UastFormat) *C.Uast {
+	if format == 0 {
+		format = C.UAST_BINARY
+	}
+	data := C.GoBytes(p, C.int(sz))
+
+	nd := &goNodes{}
+
+	c := newContext(nd)
+	u := newUast(goImpl, c.Handle())
+
+	var (
+		n   nodes.Node
+		err error
+	)
+	switch format {
+	case C.UAST_BINARY:
+		n, err = nodesproto.ReadTree(bytes.NewReader(data))
+	case C.UAST_YAML:
+		n, err = uastyml.Unmarshal(data)
+	default:
+		err = fmt.Errorf("unknown format: %v", format)
+	}
+	if err != nil {
+		c.last = err
+		return u
+	}
+	if n != nil {
+		u.root = C.NodeHandle(nd.toNode(n).Handle())
+	}
+	u.ctx = C.UastHandle(c.Handle())
+	return u
+}
+
+//export UastEncode
+func UastEncode(ctx *C.Uast, node C.NodeHandle, size *C.size_t, format C.UastFormat) unsafe.Pointer {
+	if format == 0 {
+		format = C.UAST_BINARY
+	}
+	c := getContextFrom(ctx)
+	if c == nil {
+		return nil
+	}
+	if node == 0 {
+		node = ctx.root
+	}
+
+	h := Handle(node)
+	root := c.impl.AsNode(h)
+	n, err := loadNode(root)
+	if err != nil {
+		c.setError(err)
+		return nil
+	}
+
+	buf := bytes.NewBuffer(nil)
+	switch format {
+	case C.UAST_BINARY:
+		err = nodesproto.WriteTo(buf, n)
+	case C.UAST_YAML:
+		err = uastyml.NewEncoder(buf).Encode(n)
+	default:
+		err = fmt.Errorf("unknown format: %v", format)
+	}
+	if err != nil {
+		c.setError(err)
+		return nil
+	}
+	sz := buf.Len()
+	if size != nil {
+		*size = C.size_t(sz)
+	}
+	return C.CBytes(buf.Bytes())
+}
+
+//export UastFree
+// Releases Uast resources.
+func UastFree(ctx *C.Uast) {
+	if ctx == nil {
+		return
+	}
+	h := Handle(ctx.handle)
+	C.free(unsafe.Pointer(ctx))
+	getContext(h).free()
```

Comment:
You are right in general, but the `free` method is defined in a following way:
```golang
func (c *Context) free() {
	if c == nil {
		return
	}
```
So it will check for nil pointer instead of the caller.